### 일반 크롤링
- 나무위키 
    - SF 영화, 아이돌, 암호화폐, 마카오, 프로그래밍 언어 를 크롤링
        - https://namu.wiki/w/%EB%B6%84%EB%A5%98:SF%20%EC%98%81%ED%99%94
        - https://namu.wiki/w/%EB%B6%84%EB%A5%98:%EC%95%84%EC%9D%B4%EB%8F%8C
        - https://namu.wiki/w/%EB%B6%84%EB%A5%98:%EC%95%94%ED%98%B8%ED%99%94%ED%8F%90
        - https://namu.wiki/w/%EB%B6%84%EB%A5%98:%EB%A7%88%EC%B9%B4%EC%98%A4
        - https://namu.wiki/w/%EB%B6%84%EB%A5%98:%ED%94%84%EB%A1%9C%EA%B7%B8%EB%9E%98%EB%B0%8D%20%EC%96%B8%EC%96%B4

#### 필요 라이브러리 임포트

In [3]:
from fake_useragent import UserAgent                # 페이지 요청 시 가상 유저 만들기
from time import sleep
from multiprocessing import Pool                    # 멀티프로세싱 라이브러리
from concurrent.futures import ThreadPoolExecutor   # 스레드를 프로세스로 전환/역전환
from bs4 import BeautifulSoup                       # 웹페이지 스크래핑
from selenium import webdriver                      # 웹페이지 동적 웹크롤링 
import concurrent.futures
import urllib.request  
import requests
import time

#### 크롤링할 URL리스트

In [31]:
urls = [f'https://namu.wiki/w/%EB%B6%84%EB%A5%98:SF%20%EC%98%81%ED%99%94', 
        f'https://namu.wiki/w/%EB%B6%84%EB%A5%98:%EC%95%84%EC%9D%B4%EB%8F%8C', 
        f'https://namu.wiki/w/%EB%B6%84%EB%A5%98:%EC%95%94%ED%98%B8%ED%99%94%ED%8F%90', 
        f'https://namu.wiki/w/%EB%B6%84%EB%A5%98:%EB%A7%88%EC%B9%B4%EC%98%A4',
        f'https://namu.wiki/w/%EB%B6%84%EB%A5%98:%ED%94%84%EB%A1%9C%EA%B7%B8%EB%9E%98%EB%B0%8D%20%EC%96%B8%EC%96%B4']

In [4]:
urls = [f'https://namu.wiki/w/%EB%B6%84%EB%A5%98:SF%20%EC%98%81%ED%99%94', 
        f'https://namu.wiki/w/%EB%B6%84%EB%A5%98:%EC%95%84%EC%9D%B4%EB%8F%8C', 
        f'https://namu.wiki/w/%EB%B6%84%EB%A5%98:%EC%95%94%ED%98%B8%ED%99%94%ED%8F%90']

In [5]:
## 전체 URL에서 서브링크 가져오기 함수
def get_sublist_href(url: str):
    res_links = []
    request = requests.get(url)

    parsed_html = BeautifulSoup(request.text, 'html.parser')
    a_tags = parsed_html.find_all('div', attrs={'class' : 'UvUHYDlr'})
    for tag in a_tags:
        for link in tag.find_all('a'):
            res_links.append(url + link['href'])

    print(f'Number of site: {len(res_links)}')
    return res_links

## 각 페이지 크롤링 함수
def do_html_crawl(url: str):
    request = requests.get(url)
    parsed_html = BeautifulSoup(request.text, 'html.parser')
    return parsed_html

#### 크롤링 시작

In [32]:
start_time = time.time()
for url in urls:
    sub_url_list = get_sublist_href(url)
    # print(sub_url_list)
    for sub_url in sub_url_list:
        do_html_crawl(sub_url)

time_lapse = time.time() - start_time
print(f'Total time {time_lapse} seconds.')

Number of site: 0
Number of site: 0
Number of site: 0
Number of site: 0
Number of site: 0
Total time 0.5982675552368164 seconds.


#### Nested tqdm으로 프로그래스바 표시

In [6]:
from tqdm.notebook import trange, tqdm

In [5]:
for i in trange(3, desc='1st loop'):
    for j in tqdm(range(100), desc='2nd loop' , leave = False):
        sleep(0.01)

1st loop:   0%|          | 0/3 [00:00<?, ?it/s]

2nd loop:   0%|          | 0/100 [00:00<?, ?it/s]

2nd loop:   0%|          | 0/100 [00:00<?, ?it/s]

2nd loop:   0%|          | 0/100 [00:00<?, ?it/s]

In [9]:
%pip install --upgrade jupyter ipywidgets

   ---------------------------------------- 0.0/139.8 kB ? eta -:--:--
   ---------------------------------------- 139.8/139.8 kB 8.1 MB/s eta 0:00:00
   ---------------------------------------- 0.0/214.4 kB ? eta -:--:--
   ---------------------------------------- 214.4/214.4 kB 6.4 MB/s eta 0:00:00
   ---------------------------------------- 0.0/2.3 MB ? eta -:--:--
   --------------------- ------------------ 1.2/2.3 MB 38.1 MB/s eta 0:00:01
   ------------------------- -------------- 1.5/2.3 MB 15.7 MB/s eta 0:00:01
   ---------------------------------- ----- 2.0/2.3 MB 14.3 MB/s eta 0:00:01
   ---------------------------------------- 2.3/2.3 MB 14.8 MB/s eta 0:00:00
   ---------------------------------------- 0.0/11.6 MB ? eta -:--:--
   ----- ---------------------------------- 1.5/11.6 MB 47.7 MB/s eta 0:00:01
   --------- ------------------------------ 2.9/11.6 MB 36.0 MB/s eta 0:00:01
   --------------- ------------------------ 4.5/11.6 MB 36.2 MB/s eta 0:00:01
   --------------

  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
  Consider adding this directory to PATH or, if you prefer to suppress t

In [13]:
!pip install --upgrade jupyter_client


[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [19]:
!pip show ipywidgets

Name: ipywidgets
Version: 8.1.5
Summary: Jupyter interactive widgets
Home-page: http://jupyter.org
Author: Jupyter Development Team
Author-email: jupyter@googlegroups.com
License: BSD 3-Clause License
Location: C:\Users\perso\.pyenv\pyenv-win\versions\3.11.9\Lib\site-packages
Requires: comm, ipython, jupyterlab-widgets, traitlets, widgetsnbextension
Required-by: jupyter


In [20]:
!pip install --upgrade ipywidgets


[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [6]:
# 이건 안됨
!jupyter nbextension enable --py widgetsnbextension

usage: jupyter [-h] [--version] [--config-dir] [--data-dir] [--runtime-dir]
               [--paths] [--json] [--debug]
               [subcommand]

Jupyter: Interactive Computing

positional arguments:
  subcommand     the subcommand to launch

options:
  -h, --help     show this help message and exit
  --version      show the versions of core jupyter packages and exit
  --config-dir   show Jupyter config dir
  --data-dir     show Jupyter data dir
  --runtime-dir  show Jupyter runtime dir
  --paths        show all Jupyter paths. Add --json for machine-readable
                 format.
  --json         output paths as machine-readable json
  --debug        output debug information about paths

Available subcommands: console contrib dejavu events execute kernel kernelspec
lab labextension labhub migrate nbconvert nbextensions_configurator notebook
run server troubleshoot trust

Jupyter command `jupyter-nbextension` not found.


In [24]:
!pip install jupyter_contrib_nbextensions

     ---------------------------------------- 0.0/23.5 MB ? eta -:--:--
     --------------------------------------- 0.0/23.5 MB 330.3 kB/s eta 0:01:11
     --------------------------------------- 0.1/23.5 MB 656.4 kB/s eta 0:00:36
      --------------------------------------- 0.4/23.5 MB 2.9 MB/s eta 0:00:08
     - -------------------------------------- 0.9/23.5 MB 5.3 MB/s eta 0:00:05
     -- ------------------------------------- 1.5/23.5 MB 6.8 MB/s eta 0:00:04
     --- ------------------------------------ 2.2/23.5 MB 8.1 MB/s eta 0:00:03
     ---- ----------------------------------- 2.9/23.5 MB 9.3 MB/s eta 0:00:03
     ------ --------------------------------- 3.8/23.5 MB 10.4 MB/s eta 0:00:02
     ------- -------------------------------- 4.6/23.5 MB 11.3 MB/s eta 0:00:02
     --------- ------------------------------ 5.8/23.5 MB 12.7 MB/s eta 0:00:02
     ----------- ---------------------------- 6.9/23.5 MB 13.7 MB/s eta 0:00:02
     ------------- -------------------------- 8.2/23.


[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [25]:
!jupyter contrib nbextension install --user

Traceback (most recent call last):
  File "C:\Users\perso\.pyenv\pyenv-win\versions\3.11.9\Lib\site-packages\jupyter_contrib_core\notebook_compat\nbextensions.py", line 6, in <module>
    from notebook.extensions import BaseExtensionApp
ModuleNotFoundError: No module named 'notebook.extensions'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\perso\.pyenv\pyenv-win\versions\3.11.9\Lib\site-packages\jupyter_contrib_core\notebook_compat\nbextensions.py", line 10, in <module>
    from notebook.nbextensions import BaseNBExtensionApp
ModuleNotFoundError: No module named 'notebook.nbextensions'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\perso\.pyenv\pyenv-win\versions\3.11.9\Scripts\jupyter-contrib.EXE\__main__.py", line 7, in <module>
  F

In [34]:
for i in trange(3, desc='1st loop'):
    for j in tqdm(range(100), desc='2nd loop' , leave = False):
        sleep(0.01)

1st loop:   0%|          | 0/3 [00:00<?, ?it/s]

2nd loop:   0%|          | 0/100 [00:00<?, ?it/s]

2nd loop:   0%|          | 0/100 [00:00<?, ?it/s]

2nd loop:   0%|          | 0/100 [00:00<?, ?it/s]

In [7]:
start_time = time.time()
for url in tqdm(urls, desc='outer loop'):
    sub_url_list = get_sublist_href(url)
    # print(url)
    for sub_url in tqdm(sub_url_list, desc='inner loop', leave=False):
        # print(sub_url)
        do_html_crawl(sub_url)

time_lapse = time.time() - start_time
print(f'Total time {time_lapse} seconds.')

outer loop:   0%|          | 0/3 [00:00<?, ?it/s]

Number of site: 36


inner loop:   0%|          | 0/36 [00:00<?, ?it/s]

Number of site: 64


inner loop:   0%|          | 0/64 [00:00<?, ?it/s]

Number of site: 127


inner loop:   0%|          | 0/127 [00:00<?, ?it/s]

Total time 35.887481451034546 seconds.


#### 총 59초 소요

#### 세가지 링크로 진행했을 때는 35.9초 소요

### 멀티스레드 크롤링

In [8]:
## 스레드 크롤링
def do_thread_crawl(urls: list):
    thread_list = []
    with ThreadPoolExecutor(max_workers=12) as executor:
        for url in urls:
            thread_list.append(executor.submit(do_html_crawl, url)) ## 실행시 함수와 파라미터를 submit()에 나열
        for execution in concurrent.futures.as_completed(thread_list):
            execution.result()

#### 크롤링 시작

In [9]:
start_time = time.time()
for url in tqdm(urls, desc='one loop'):
    do_thread_crawl(get_sublist_href(url))

time_lapse = time.time() - start_time
print(f'Total time {time_lapse} seconds.')

one loop:   0%|          | 0/3 [00:00<?, ?it/s]

Number of site: 36
Number of site: 64
Number of site: 127
Total time 5.411362648010254 seconds.


#### 스레드워커를 12개 까지 올렸을 경우, 4.4초 밖에 걸리지 않음

#### 나무위키 비정상적 트래픽 처리로 네다섯번 리스트에서 처리 안됨
- 세개 항목으로 변경!

### 멀티프로세스 처리

In [38]:
def do_process_with_thread_crawl(url: str):
    do_thread_crawl(get_sublist_href(url))

In [ ]:
start_time = time.time()

with Pool(processes=4) as pool:  
    pool.map(do_process_with_thread_crawl, urls)

time_lapse = time.time() - start_time
print(f'Total time {time_lapse} seconds.')

#### 멀티프로세스에서 무조건 속도가 줄어들지는 않음

In [1]:
!pip install ray

   ---------------------------------------- 0.0/24.9 MB ? eta -:--:--
    --------------------------------------- 0.5/24.9 MB 9.4 MB/s eta 0:00:03
   -- ------------------------------------- 1.8/24.9 MB 18.8 MB/s eta 0:00:02
   ------ --------------------------------- 4.3/24.9 MB 30.4 MB/s eta 0:00:01
   ----------- ---------------------------- 7.2/24.9 MB 38.2 MB/s eta 0:00:01
   ------------ --------------------------- 7.5/24.9 MB 37.0 MB/s eta 0:00:01
   ------------ --------------------------- 7.6/24.9 MB 28.6 MB/s eta 0:00:01
   ------------ --------------------------- 7.8/24.9 MB 24.8 MB/s eta 0:00:01
   ------------ --------------------------- 7.9/24.9 MB 22.0 MB/s eta 0:00:01
   ------------- -------------------------- 8.6/24.9 MB 21.1 MB/s eta 0:00:01
   ------------------ --------------------- 11.5/24.9 MB 26.2 MB/s eta 0:00:01
   ---------------------- ----------------- 14.3/24.9 MB 28.5 MB/s eta 0:00:01
   --------------------------- ------------ 17.1/24.9 MB 27.3 MB/s eta 


[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [11]:
!pip --version

pip 24.2 from C:\Users\perso\.pyenv\pyenv-win\versions\3.11.9\Lib\site-packages\pip (python 3.11)



#### Ray 병렬처리 미사용

In [2]:
def square(number):
    return number * number

In [3]:
futures = [square(i) for i in range(1, 10000001)]

In [4]:
print(len(futures))

10000000


#### Ray 사용

In [2]:
import ray

In [3]:
# Ray 초기화
ray.init()

2024-10-21 16:00:26,380	INFO worker.py:1786 -- Started a local Ray instance.


RayContext(dashboard_url='', python_version='3.11.9', ray_version='2.37.0', ray_commit='1b620f244c4fb67d80661046091573fb6eefecb6')

In [7]:
# 병렬로 실행할 함수 정의
@ray.remote
def square(number):
    return number * number

In [9]:
# 병렬로 1부터 100까지의 제곱을 계산
futures = [square.remote(i) for i in range(1, 10000001)]

KeyboardInterrupt: 

In [11]:
!pip install opencv-python

   ---------------------------------------- 0.0/38.8 MB ? eta -:--:--
   ------ --------------------------------- 6.6/38.8 MB 40.1 MB/s eta 0:00:01
   ---------------- ----------------------- 16.0/38.8 MB 41.9 MB/s eta 0:00:01
   ------------------------ --------------- 24.1/38.8 MB 41.2 MB/s eta 0:00:01
   --------------------------------- ------ 32.5/38.8 MB 40.5 MB/s eta 0:00:01
   ---------------------------------------  38.8/38.8 MB 41.1 MB/s eta 0:00:01
   ---------------------------------------- 38.8/38.8 MB 38.0 MB/s eta 0:00:00


In [12]:
!pip install opencv-contrib-python

   ---------------------------------------- 0.0/45.5 MB ? eta -:--:--
   ------- -------------------------------- 8.1/45.5 MB 42.1 MB/s eta 0:00:01
   ----------------- ---------------------- 19.9/45.5 MB 50.4 MB/s eta 0:00:01
   -------------------------- ------------- 29.6/45.5 MB 48.2 MB/s eta 0:00:01
   ------------------------------- -------- 35.7/45.5 MB 42.7 MB/s eta 0:00:01
   ---------------------------------------  45.4/45.5 MB 43.7 MB/s eta 0:00:01
   ---------------------------------------- 45.5/45.5 MB 40.2 MB/s eta 0:00:00


In [14]:
import numpy as np
import time
import cv2

num_cpus = 16

filters = [np.random.normal(size=(4, 4)) for _ in range(num_cpus)]
aver_ptime = 0
for _ in range(10):
    start = time.time()
    image = np.zeros((3000, 3000))
    result = []
    for i in range(num_cpus):
        result.append(cv2.filter2D(image, -1, filters[i]))
    ptime = time.time() - start
    aver_ptime += ptime / 10
    print(f'{ptime:.3f}s')
print(f'aver ptime : {aver_ptime:.3f}s')

1.072s
1.159s
1.164s
1.141s
1.102s
1.185s
1.096s
1.142s
1.126s
1.268s
aver ptime : 1.146s


In [15]:
import numpy as np
import time
import cv2
import ray

num_cpus = 16
ray.init(num_cpus = num_cpus, ignore_reinit_error=True)

@ray.remote
def f(image, kernel):
    return cv2.filter2D(image, -1, kernel)

filters = [np.random.normal(size=(4, 4)) for _ in range(num_cpus)]

aver_ptime = 0
for _ in range(10):
    start = time.time()
    image = np.zeros((3000, 3000))
    image_id = ray.put(image)
    result = ray.get([f.remote(image_id, filters[i]) for i in range(num_cpus)])
    ptime = time.time() - start
    aver_ptime += ptime / 10
    print(f'{ptime:.3f}s')
print(f'aver ptime : {aver_ptime:.3f}s')

2024-10-21 15:56:51,980	INFO worker.py:1619 -- Calling ray.init() again after it has already been called.


1.042s
0.572s
0.573s
0.438s
0.542s
0.534s
0.568s
0.544s
0.431s
0.550s
aver ptime : 0.579s


In [ ]:
import numpy as np
import cv2
from multiprocessing import Pool

num_cpus = 16

def edge_multi(args):
    image, kernel = args
    return cv2.filter2D(image, -1, kernel)

pool = Pool(num_cpus)

filters = [np.random.normal(size=(4, 4)) for _ in range(num_cpus)]

aver_ptime = 0
for _ in range(10):
    start = time.time()
    image = np.zeros((3000, 3000))
    pool.map(edge_multi, zip(num_cpus * [image], filters))
    ptime = time.time() - start
    aver_ptime += ptime / 10
    print(f'{ptime:.3f}s')
print(f'aver ptime : {aver_ptime:.3f}s')